enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,298 @@ package hit_tracker_client
-	return &HitTracker{}
+type NoOpTransferTimer struct {
 }
 
-func (h HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+func (t *NoOpTransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
+	return nil
+}
+
+func (t *NoOpTransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
+	return nil
 }
 
-func (h *HitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+type HitTrackerClient struct {
+	ctx             context.Context
+	enqueueFn       func(context.Context, *repb.RequestMetadata, *hitpb.CacheHit)
+	client          hitpb.HitTrackerServiceClient
+	requestMetadata *repb.RequestMetadata
 }
 
-func (h *HitTracker) TrackMiss(d *repb.Digest) error {
+// This should only be used for Action Cache hit tracking.
+func (h *HitTrackerClient) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+	alert.UnexpectedEvent("Unexpected call to SetExecutedActionMetadata")
+}
+
+// Misses hit the backing cache, so no need to report on hit-tracking.
+func (h *HitTrackerClient) TrackMiss(d *repb.Digest) error {
 	return nil
 }
 
+func (h *HitTrackerFactory) groupID(ctx context.Context) groupID {
+	user, err := h.authenticator.AuthenticatedUser(ctx)
+	if err != nil {
+		return interfaces.AuthAnonymousUser
+	}
+	return groupID(user.GetGroupID())
+}
+
+func (h *HitTrackerFactory) enqueue(ctx context.Context, requestMetadata *repb.RequestMetadata, hit *hitpb.CacheHit) {
+	if h.stopped.Load() {
+		log.Warning("hit_tracker_client.enqueue after worker shutdown, sending RPC synchronously")
+		hit.RequestMetadata = requestMetadata
+		if _, err := h.client.Track(ctx, &hitpb.TrackRequest{Hits: []*hitpb.CacheHit{hit}}); err != nil {
+			log.Infof("Error sending HitTrackerService.Track RPC: %v", err)
+		}
+		return
+	}
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	groupID := h.groupID(ctx)
+	if _, ok := h.hitsByGroup[groupID]; !ok {
+		h.hitsByGroup[groupID] = &cacheHits{
+			hits: make([]*hitpb.CacheHit, 0, initialSliceCapacity),
+		}
+	}
+
+	groupHits := h.hitsByGroup[groupID]
+	if len(groupHits.hits) >= h.maxPendingHitsPerGroup {
+		log.CtxWarning(ctx, "Exceeded maximum number of pending cache hits, dropping some.")
+		metrics.RemoteHitTrackerUpdates.With(
+			prometheus.Labels{
+				metrics.GroupID:              string(groupID),
+				metrics.EnqueueUpdateOutcome: "dropped_too_many_updates",
+			}).Add(float64(1))
+		return
+	}
+	metrics.RemoteHitTrackerUpdates.With(
+		prometheus.Labels{
+			metrics.GroupID:              string(groupID),
+			metrics.EnqueueUpdateOutcome: "enqueued",
+		}).Add(float64(1))
+
+	authHeaders := authutil.GetAuthHeaders(ctx, h.authenticator)
+	groupHits.authHeaders = authHeaders
```

Comment:
nit: add a comment that we're always setting the latest auth headers on the groupHits (or make this conditional on them being unset if that's what you want -- maybe that is less work since we don't have to pull auth stuff?)

Reply:
Added a comment